rfcs/20200619-dropout/README.md

```diff
@@ -0,0 +1,425 @@
+        yield seed
+```
+
+In VSL MCG31 engine is implemented slightly different - it skips "c" parameter
+(consulted with MKL team). With this change, the core of the formula simplifies
+to:
+
+```python
+seed = (a * seed) % modulus
+```
+
+It is worth noting that in such case:
+
+```python
+seed[1] = seed[0] * A % modulus
+seed[2] = seed[0] * (A * A % modulus) % modulus = seed[1] * A % modulus
+```
+
+So, first N random values can be computed as a multiplication of vector of
+initial seeds (scalar seed broadcasted to vector) by vector of consecutive powers
+of "a" modulo "modulus".
+The latter vectors of random values, as they depend on the former ones, can be
+computed by vector multiplication of previously generated output seed and a
+broadcasted power of a modulo "modulus".
+
+With that said, we can introduce a following vectorized formula for computing
+vectors of random values (with utilization of SIMD instructions):
+
+```
+N - lenght of AVX vector
+
+// first iteration (i=0)
+seed_t[0:N] = [seed, seed, ..., seed]
+a_t[0:N] = [1, A % modulus, A^2 % modulus, ..., A^N % modulus]
+out[0:N] = seed[0:N] * a_t[0:N] % modulus
+
+//following iterations
+aa_t[0:N] = [A^N % modulus, A^N % modulus, ..., A^N % modulus]
+for(i=1:...)
+    out[N*i:N*(i+1)] = out[N*(i-1):N*i] * aa_t[0:N] % modulus
+```
+
+According to GPU team opinion (analysis made by Chereshnev, Eugene) the above
+algorithm should be fairly easily portable to GPU.
+
+### RNG State
+
+Should we store RNG state somewhere ? oneDNN primitives should be stateless.
+
+Possible use case:
+```cpp
+primitve rng(...);
+rng.execute(...) ;
+rng.execute(...) // second exeucte should start from end state of previous execute;
+```
+Let's consider https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/dropout-inl.h
+since line 90. In each time they are creating new stream vsl stream and use
+viRngBernoulli. They don't save stream and viRngBernoulli doesn't save any state.
+The only thing that they're doing is gnenerating new seed in each param.
+
+Possible solution:
+- User should pass seed in execute instead primitive_descriptor (in intial draft,
+seed was passed in the primitive descriptor)
+- User passes seed in execute using DNNL_ARG_SEED. Primitivie in
+DNNL_ARG_SEED memory returns end-state of rng engine. In next calls of 'execute'
+
+## Open questions
+
+1. Should user be allowed to specify noise shape / axes  ?
+
+In our opinion yes. It is used in two popular frameworks - tensorflow, mxnet.
+Rationale for that is further decrease of memory usage + if we ommit now that
+parameter and if in the future there will be request to support that, it will
+require either add some kind v2 version in api or abi break.
+
+2. Should we support inplace computation (pytorch) ?
+
+In our opionion we can support that, it is not big implementation effort.
+
+3. Do we understand how differences in RNGs affect training results ?
+
+4. Where in DNN is RNG except dropout ?
```

Comment:
https://mxnet.cdn.apache.org/api/python/docs/api/ndarray/random/index.html#module-mxnet.ndarray.random

Reply:
Thank you for answer :-) Maybe I missed something from you reply or question is not well formed. I mainly meant in what real deep neural network topologies random generators are used, in particular in what popular primitives. It is quite obvious that RNG it is used in various random like functions and distribution generators. I wanted to know practical usage of different frameworks random modules in applications other than dropout. I received already answer from frameworks team and I will update RFC in further commit.